File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Native/AlxTopOnNativeAdapter.m
//
//  AlxTopOnNativeAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnNativeAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnNativeDelegate.h"
#import "AlxTopOnNativeEvent.h"
#import "AlxTopOnNativeObject.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"
#import "AlxTopOnNativeRender.h"

static NSString *const TAG = @"AlxTopOnNativeAdapter";

@interface AlxTopOnNativeAdapter ()
@property (nonatomic, strong) AlxNativeAdLoader *nativeAdLoader;
@property (nonatomic, strong) AlxNativeAd *nativeAd;
@property (nonatomic, strong) AlxTopOnNativeDelegate *nativeDelegate;
/**
 * 用于传递给 TopOn SDK。
 * For passing to TopOn SDK.
 */
@property (nonatomic, strong) AlxTopOnNativeEvent *nativeEvent;
@end

@implementation AlxTopOnNativeAdapter

#pragma mark - lazy load
- (AlxTopOnNativeDelegate *)nativeDelegate {
    if (_nativeDelegate == nil) {
        _nativeDelegate = [[AlxTopOnNativeDelegate alloc] init];
        _nativeDelegate.adStatusBridge = self.adStatusBridge;
        _nativeDelegate.nativeEvent = self.nativeEvent;  // 设置 Event 引用 / Set Event reference
    }
    return _nativeDelegate;
}

- (AlxTopOnNativeEvent *)nativeEvent {
    if (_nativeEvent == nil) {
        // 创建一个 Event 对象用于传递给 TopOn SDK
        // Create an Event object for passing to TopOn SDK
        _nativeEvent = [[AlxTopOnNativeEvent alloc] initWithInfo:@{} localInfo:@{}];
    }
    return _nativeEvent;
}

#pragma mark - Ad load
- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadADWithArgument", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
    
    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *unitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *unitId = argument.serverContentDic[unitKeyStr];
        
        if (!unitId || unitId.length == 0) {
            NSString *errorStr = @"unitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            return;
        }
        NSLog(@"%@: loadAD: unitid = %@", TAG, unitId);
        
        if (bidId) {
            // Bidding 场景：从缓存中取出已加载的广告
            // Bidding scenario: retrieve the pre-loaded ad from cache
            AlxTopOnBiddingRequest *biddingRequest = [[AlxTopOnTool shared] getRequestItemWithUnitID:unitId];
            if (biddingRequest) {
                self.nativeAd = (AlxNativeAd *)biddingRequest.customObject;
                
                if (self.nativeAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, creating native object", TAG);
                    
                    // ⚠️ 正确方式：创建 AlxTopOnNativeObject 对象
                    // ⚠️ Correct approach: create an AlxTopOnNativeObject instance
                    AlxTopOnNativeObject *nativeObject = [[AlxTopOnNativeObject alloc] init];
                    nativeObject.nativeAd = self.nativeAd;
                    nativeObject.nativeEvent = self.nativeEvent;
                    
                    // ✅ 关键：设置 nativeAd 的 delegate，以便接收展示、点击、关闭回调
                    // ✅ Key: set nativeAd's delegate to receive impression, click, and close callbacks
                    self.nativeAd.delegate = self.nativeDelegate;
                    
                    // ⚠️ 传递对象数组 / ⚠️ Pass the object array
                    [self.adStatusBridge atOnNativeAdLoadedArray:@[nativeObject] adExtra:@{}];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            }
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:unitId];
        } else {
            // 普通加载场景 / Normal loading scenario
            self.nativeAdLoader = [[AlxNativeAdLoader alloc] initWithAdUnitID:unitId];
            self.nativeAdLoader.delegate = self.nativeDelegate;
            
            NSLog(@"%@: start loading ad with unitId: %@", TAG, unitId);
            [self.nativeAdLoader loadAd];
        }
    });
}


#pragma mark - C2S Bidding
+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });
    
    AlxTopOnNativeEvent *customEvent = [[AlxTopOnNativeEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;
    
    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatNative
                                       bidCompletion:completion];
    
    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

/**
 * 实现协议中的类方法。
 * Implement the class method defined in the protocol.
 */
+ (Class)rendererClass {
    NSLog(@"%@: rendererClass", TAG);
    // 返回对应的渲染类 / Return the corresponding renderer class
    return [AlxTopOnNativeRender class];
}

@end

File: custom_adapter/TopOnAdapter/AlxTopOnNativeAdapter.swift
//
//  AlxTopOnNativeAdapter.swift
//

import Foundation
import AnyThinkSDK
import AnyThinkNative
import AlxAds

@objc(AlxTopOnNativeAdapter)
public class AlxTopOnNativeAdapter:ATAdAdapter {
    
    private static let TAG = "AlxTopOnNativeAdapter"
    
    private var nativeAd:AlxNativeAdLoader? = nil
    private var customEvent:AlxTopOnNativeEvent? = nil
    
    
    // MARK: - Initialization
    @objc public required init(networkCustomInfo serverInfo: [AnyHashable: Any],
                               localInfo: [AnyHashable: Any]) {
        super.init()
        NSLog("%@: init",AlxTopOnNativeAdapter.TAG)
        NSLog("%@: init: isMainThread=%@",AlxTopOnNativeAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
        }
    }
    
    /// 加载广告
    @objc public func loadAD(withInfo serverInfo: [AnyHashable: Any],
                             localInfo: [AnyHashable: Any],
                             completion: @escaping ([[AnyHashable: Any]]?, (any Error)?) -> Void) {
        NSLog("%@: loadAD",AlxTopOnNativeAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@",AlxTopOnNativeAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
            
            let bidId=serverInfo[kATAdapterCustomInfoBuyeruIdKey] as? String
            NSLog("%@: loadAD: bidId=%@",AlxTopOnNativeAdapter.TAG,bidId ?? "")     
            
            guard let unitId = serverInfo[AlxTopOnBaseManager.unitID] as? String,!unitId.isEmpty else{
                let errorStr="unitid is empty"
                NSLog("%@: loadAD: error= %@",AlxTopOnNativeAdapter.TAG,errorStr)
                completion(nil,AlxTopOnBaseManager.error(code: -100, msg: errorStr))
                return
            }
            
            NSLog("%@: loadAD: unitid=%@",AlxTopOnNativeAdapter.TAG,unitId)
            
            if bidId != nil {
                let request:AlxTopOnBiddingRequest? = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest
                if let request = request {
                    self.customEvent = request.customEvent as? AlxTopOnNativeEvent
                    self.customEvent?.requestCompletionBlock = completion
                    self.nativeAd = request.customObject as? AlxNativeAdLoader
                    //判断广告源是否已经loaded过
                    if let assets = self.customEvent?.assetDict {
                        NSLog("%@: loadAD: bid load success callback",AlxTopOnNativeAdapter.TAG)
                        self.customEvent?.trackNativeAdLoaded([assets])
                    }else{
                        NSLog("%@: loadAD: bid ad object is empty",AlxTopOnNativeAdapter.TAG)
                    }
                    self.customEvent?.assets.removeAllObjects()
                    self.customEvent?.assetDict = [:]
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            }else{
                
                self.customEvent = AlxTopOnNativeEvent(info: serverInfo, localInfo: localInfo)
                self.customEvent?.requestCompletionBlock = completion
                
                // load ad
                self.nativeAd = AlxNativeAdLoader(adUnitID: unitId)
                self.nativeAd?.delegate = self.customEvent
                self.nativeAd?.loadAd()
            }
        }
        
    }
    
    // MARK: - C2S header bidding 竞价
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,unitGroupModel: ATUnitGroupModel, info: [AnyHashable: Any], completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel",AlxTopOnNativeAdapter.TAG)
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent:AlxTopOnNativeEvent = AlxTopOnNativeEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        let request:AlxTopOnBiddingRequest = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.native,
            bidCompletion: completion
        )
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
        
    // 实现 Objective-C 协议中的类方法: + (Class)rendererClass;
    @objc public static func rendererClass() -> AnyClass {
        NSLog("%@: rendererClass",AlxTopOnNativeAdapter.TAG)
        return AlxTopOnNativeRender.self
    }
    
}
